### PR TITLE
Add Jenkinsfile to allow create contrail from jenkins pipeline

### PR DESCRIPTION
**Pull request by iosetek**

### CODE REVIEW

File: provision/kubernetes/kubernetes.tf [hunk 1]
```diff
@@ -39,122 +32,22 @@ resource "openstack_compute_secgroup_v2" "contrail_security_group" {
...
-    to_port     = 6080
-    ip_protocol = "tcp"
-    cidr        = "0.0.0.0/0"
-  }
+resource "openstack_blockstorage_volume_v2" "volume" {
+  name = "${var.user_name}-volume"
+  size = 100
 }
 
 resource "openstack_compute_instance_v2" "basic" {
```
Comment (kravvcuCL) [inline provision/kubernetes/kubernetes.tf:47]: can you rename this resource? It does not say anything about what the instance is for. Same for the volume + same for openstack.tf. Though, you'll have to rename the resources in the tfstate files also :-3

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                        sh "rm -rf ${sshPrivKeyFile}"
+                        sh "mv ${instancesYaml} provision/template.yaml"
+                        sh "chmod 600 ${params.Login}-key.pub"
+                        sh "chmod 600 ${params.Login}-key.priv"
+                        sh "chmod 777 provision/prepare_template"
+                        // set +x and set -x are workaround to not print user password in jenkins output log
+                        ansiColor('xterm') {
+                            sh "set +x && cd provision && terraform init -from-module ${params.orchestrator} && ./createcontrail \"${params.CreateDestroy}\" \"${params.Login}\" \"${params.Password}\" \"${params.ProjectID}\" \"${params.domainName}\" \"${params.projectName}\" \"${params.networkName}\" \"../${params.Login}-key.pub\" \"../${params.Login}-key.priv\" \"${params.routerIP}\" \"${params.orchestrator}\" \"${params.branch}\" && set -x"
+                        }
+                        sh "rm -rf ${params.Login}-key.pub"
```
Comment (kravvcuCL) [inline Jenkinsfile]: Will this execute if the terraform command fails? I'd put this in a 'finally' or similar (if available) Jenkins pipeline phase.

File: vars/unstashParam.groovy [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import hudson.FilePath
+import hudson.model.ParametersAction
+import hudson.model.FileParameterValue
+import hudson.model.Executor
+
+def call(String name, String fname = null) {
+    def paramsAction = currentBuild.rawBuild.getAction(ParametersAction.class);
+    if (paramsAction != null) {
+        for (param in paramsAction.getParameters()) {
```
Comment (kravvcuCL) [inline vars/unstashParam.groovy:12]: seems like a good case of using streams but I'm leaving it up to you if you want to change it http://mrhaki.blogspot.com/2015/04/groovy-goodness-use-closures-as-java.html

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                    checkout scm
+                }
+                stash name: "Provision", includes: "provision/**"
+                script {
+                    unstash "Provision"
+                    if ("${sshpubkey}" != "" && "${sshprivkey}" != "") {
+                        def sshPubKeyFile = unstashParam "sshpubkey"
+                        def sshPrivKeyFile = unstashParam "sshprivkey"
+                        def instancesYaml = unstashParam "instances_yaml"
+                        sh "mv ${sshPubKeyFile} ${params.Login}-key.pub"
```
Comment (kravvcuCL) [inline Jenkinsfile]: Is this renaming necessary? It add's 4 more unneded lines I think.

File: vars/unstashParam.groovy [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+                    error "unstashParam: no node in current context"
+                }
+                if (env['WORKSPACE'] == null) {
+                    error "unstashParam: no workspace in current context"
+                }
+                nodeName = env['NODE_NAME'] == 'master' ? '(master)' : env['NODE_NAME']
+                workspace = new FilePath(Jenkins.getInstance().getComputer(nodeName).getChannel(), env['WORKSPACE'])
+                filename = fname == null ? param.getOriginalFileName() : fname
+                file = workspace.child(filename)
+                file.copyFrom(param.getFile())
```
Comment (kravvcuCL) [inline vars/unstashParam.groovy:27]: Aren't the uploaded files available already in the workspace? What am I missing?

**Review by kravvcuCL — CHANGES_REQUESTED** [review]

File: provision/kubernetes/kubernetes.tf (hunk 1 above)
Comment (iosetek) [reply provision/kubernetes/kubernetes.tf:47]: But volume source exists in volume category in openstack so it is the same name.

File: Jenkinsfile (hunk 4 above)
Comment (iosetek) [reply Jenkinsfile]: It is required. Without using unstash param jenkins pipeline is not able to use file parameter correctly. It's a workaround: https://github.com/pj2/JENKINS-27413-workaround-library

File: vars/unstashParam.groovy (hunk 5 above)
Comment (iosetek) [reply vars/unstashParam.groovy:27]: Somehow files passed by file parameter are not working without this workaround:
https://github.com/pj2/JENKINS-27413-workaround-library

File: vars/unstashParam.groovy (hunk 3 above)
Comment (iosetek) [reply vars/unstashParam.groovy:12]: Let's just keep it this way

File: Jenkinsfile (hunk 2 above)
Comment (iosetek) [reply Jenkinsfile]: Done

File: vars/unstashParam.groovy (hunk 5 above)
Comment (m-kostrzewa) [reply vars/unstashParam.groovy:27]: Then please add a comment in code explaining this workaround
Not necessarily in this file, mind you, as its copied

File: Attributions.txt [hunk 6]
```diff
@@ -0,0 +1,25 @@
+vars/unstashParam.groovy
```
Comment (m-kostrzewa) [inline Attributions.txt:1]: Did you generate this file or write it by hand?

File: Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                        sh "mv ${instancesYaml} provision/template.yaml"
+                        sh "chmod 600 ${params.Login}-key.pub"
+                        sh "chmod 600 ${params.Login}-key.priv"
+                        sh "chmod 777 provision/prepare_template"
+                        // set +x and set -x are workaround to not print user password in jenkins output log
+                        ansiColor('xterm') {
+                            sh "set +x && cd provision && terraform init -from-module ${params.orchestrator} && ./createcontrail \"${params.CreateDestroy}\" \"${params.Login}\" \"${params.Password}\" \"${params.ProjectID}\" \"${params.domainName}\" \"${params.projectName}\" \"${params.networkName}\" \"../${params.Login}-key.pub\" \"../${params.Login}-key.priv\" \"${params.routerIP}\" \"${params.orchestrator}\" \"${params.branch}\" \"${params.flavor}\" && set -x"
+                        }
+                    } else {
+                        // set +x and set -x are workaround to not print user password in jenkins output log
```
Comment (m-kostrzewa) [inline Jenkinsfile]: alternatively, we could use jenkins credentials here, right? Jenkins automatically replaces them with "*".

File: provision/kubernetes/kubernetes.tf [hunk 8]
```diff
@@ -184,7 +82,7 @@ resource "openstack_compute_floatingip_associate_v2" "floatip_1" {
 
     inline = [
       "sudo yum -y install kernel-devel kernel-headers ansible git",
-      "sudo git clone http://github.com/Juniper/contrail-ansible-deployer -b ${var.branch},
+      "sudo git clone http://github.com/Juniper/contrail-ansible-deployer -b ${var.branch}",
```
Comment (m-kostrzewa) [inline provision/kubernetes/kubernetes.tf:85]: o_O why is this sudo here anyways?

File: provision/kubernetes/kubernetes.tf [hunk 9]
```diff
@@ -271,7 +168,7 @@ resource "openstack_compute_floatingip_associate_v2" "floatip_1" {
       "sudo ansible-playbook -i inventory/ -e orchestrator=kubernetes playbooks/install_k8s.yml",
       "sudo ansible-playbook -i inventory/ -e orchestrator=kubernetes playbooks/install_contrail.yml",
       "echo ${openstack_compute_instance_v2.basic.network.0.fixed_ip_v4} $HOSTNAME | sudo tee --append /etc/hosts",
-      "sudo reboot",
+      "sudo shutdown -r 1",
```
Comment (m-kostrzewa) [inline provision/kubernetes/kubernetes.tf:171]: Why this change?

File: provision/openstack/openstack.tf [hunk 10]
```diff
@@ -271,7 +163,7 @@ resource "openstack_compute_floatingip_associate_v2" "floatip_1" {
       "sudo ansible-playbook -i inventory playbooks/install_openstack.yml -v",
       "sudo ansible-playbook -i inventory/ -e orchestrator=openstack playbooks/install_contrail.yml",
       "echo ${openstack_compute_instance_v2.basic.network.0.fixed_ip_v4} $HOSTNAME | sudo tee --append /etc/hosts",
-      "sudo reboot",
+      "sudo shutdown -r 1",
```
Comment (m-kostrzewa) [inline provision/openstack/openstack.tf:166]: as above

File: vars/unstashParam.groovy (hunk 5 above)
Comment (m-kostrzewa) [reply vars/unstashParam.groovy]: Not necessarily in this file, mind you, as its copied

**Review by m-kostrzewa — APPROVED** [review]

File: Jenkinsfile (hunk 7 above)
Comment (iosetek) [reply Jenkinsfile]: We could but it would require switch function in code for every registered user in credentials.
It means updates for every new user in contrail-lab

File: Attributions.txt (hunk 6 above)
Comment (iosetek) [reply Attributions.txt:1]: Write it by my hand.

File: provision/kubernetes/kubernetes.tf (hunk 9 above)
Comment (iosetek) [reply provision/kubernetes/kubernetes.tf:171]: Otherwise console yells that there was no exit 0 called at the end and the pipeline quits with error (despite the fact it already did everything)

File: provision/kubernetes/kubernetes.tf (hunk 8 above)
Comment (iosetek) [reply provision/kubernetes/kubernetes.tf:85]: ¯\_(ツ)_/¯
I didn't notice that as it was before my commit but I can take care of it

Comment (kravvcuCL) [reply provision/kubernetes/kubernetes.tf:85]: You did not take care of it in the end? :P

File: vars/unstashParam.groovy (hunk 5 above)
Comment (kravvcuCL) [reply vars/unstashParam.groovy:27]: +1 if the code is from this workaround, please leave a note about it

File: provision/kubernetes/kubernetes.tf (hunk 1 above)
Comment (kravvcuCL) [reply provision/kubernetes/kubernetes.tf:47]: I don't understand. It's a minor though so let's leave it.

**Review by kravvcuCL — CHANGES_REQUESTED** [review]

File: provision/kubernetes/kubernetes.tf (hunk 8 above)
Comment (iosetek) [reply provision/kubernetes/kubernetes.tf:85]: It seems that every line requires sudo (if needed).
Otherwise the build cannot be completed.

File: vars/unstashParam.groovy (hunk 5 above)
Comment (iosetek) [reply vars/unstashParam.groovy:27]: Added note

File: provision/openstack/openstack.tf (hunk 10 above)
Comment (iosetek) [reply provision/openstack/openstack.tf:166]: as above:
"Otherwise console yells that there was no exit 0 called at the end and the pipeline quits with error (despite the fact it already did everything)"

**Review by kravvcuCL — APPROVED** [review]

File: provision/kubernetes/kubernetes.tf (hunk 8 above)
Comment (m-kostrzewa) [reply provision/kubernetes/kubernetes.tf:85]: Probably a permissions issue somewhere else...

**Review by kaweue — APPROVED** [review]
